### PR TITLE
(#Closes 2226) Implementation of the path_from method and corresponding test

### PR DESCRIPTION
**Pull request by LonelyCat124**

### CODE REVIEW

**Comment by LonelyCat124** [comment]
Ready for a look now I think. I'm not sure if it should be throwing a `ValueError` or another error, let me know and I'll fix it.

File: src/psyclone/psyir/nodes/node.py [hunk 1]
```diff
@@ -1502,6 +1502,32 @@ def debug_string(self):
         from psyclone.psyir.backend.debug_writer import DebugWriter
         return DebugWriter()(self)
 
+    def path_from(self, ancestor):
+        ''' Find the path in the psyir tree between ancestor and node and
+        returns a list containing the path.
```
Comment (arporter) [inline src/psyclone/psyir/nodes/node.py:1507]: Please give a bit more detail about how the resulting path can be used, e.g.:

    idx_list = node.path_from(ancestor)
    cursor = ancestor
    for idx in idx_list:
        cursor = cursor.children[idx]
    assert cursor is node

File: src/psyclone/psyir/nodes/node.py [hunk 2]
```diff
@@ -1502,6 +1502,32 @@ def debug_string(self):
...
+
+        :param ancestor: an ancestor node of self to find the path from.
+        :type ancestor: :py:class:`psyclone.psyir.nodes.Node`
+
+        :raises ValueError: if ancestor is not an ancestor of self.
+
+        :returns: a list of child indices representing the path between
+                  ancestor and self.
+        :rtype: List[Int]
```
Comment (arporter) [inline src/psyclone/psyir/nodes/node.py:1516]: s/Int/int/ I think.

File: src/psyclone/psyir/nodes/node.py [hunk 3]
```diff
@@ -1502,6 +1502,32 @@ def debug_string(self):
...
+        '''
+        result_list = []
+        current_node = self
+        while current_node is not ancestor and current_node.parent is not None:
+            result_list.append(current_node.position)
+            current_node = current_node.parent
+
+        if current_node is not ancestor:
+            raise ValueError("Attempted to find path_from a non-ancestor "
```
Comment (arporter) [inline src/psyclone/psyir/nodes/node.py:1525]: "... node ({type(node).__name__})" just to give a bit more info.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good @LonelyCat124. I did wonder whether we should/could merge the functionality into the `ancestor` method but I looked and it would be tricky due to the complexity around handling the `shared_with` argument. Therefore I'm happy to have this as a separate method.
Associated test fully covers the new method.
Changes are pylint clean.
I'll check the ref. guide next time.
Just a few tiny suggestions and then this can go on.

**Comment by arporter** [comment]
Could you also add a very brief mention of the new method to https://psyclone.readthedocs.io/en/stable/psyir.html#tree-navigation. Thanks :-)

File: src/psyclone/psyir/nodes/node.py (hunk 1 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/node.py:1507]: Added

File: src/psyclone/psyir/nodes/node.py (hunk 2 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/node.py:1516]: Fixed

File: src/psyclone/psyir/nodes/node.py (hunk 3 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/node.py]: Added

**Comment by LonelyCat124** [comment]
I think I've addressed all the changes and added a comment in the docs where you said.

File: src/psyclone/psyir/nodes/node.py [hunk 4]
```diff
@@ -1506,14 +1506,22 @@ def path_from(self, ancestor):
         ''' Find the path in the psyir tree between ancestor and node and
         returns a list containing the path.
 
+        The result of this method can be used to find the node from its
+        ancestor for example by:
+        >>> index_list = node.path_from(ancestor)
```
Comment (arporter) [inline src/psyclone/psyir/nodes/node.py:1511]: This isn't rendering correctly when I build the user guide. Perhaps you need a blank line before the first '>>>' or maybe you need to indent the '>>>' lines?

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 5]
```diff
@@ -1485,5 +1485,5 @@ def test_path_from(fortran_reader):
     loops = psyir.walk(Loop)
     with pytest.raises(ValueError) as excinfo:
         assigns[0].path_from(loops[0])
-    assert ("Attempted to find path_from a non-ancestor node."
-            in str(excinfo.value))
+    assert ("Attempted to find path_from a non-ancestor node: "
```
Comment (arporter) [inline src/psyclone/tests/psyir/nodes/node_test.py:1488]: A bit picky but how about "...non-ancestor 'Loop' node"? (i.e. move the type info and put it in quotes.)

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there. I would have merged but the additions to the docstring (which I like) are not rendering correctly.

File: src/psyclone/psyir/nodes/node.py (hunk 4 above)
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/node.py:1511]: Fixed and built locally and seems fine, just needed a blank line.

File: src/psyclone/tests/psyir/nodes/node_test.py (hunk 5 above)
Comment (LonelyCat124) [reply src/psyclone/tests/psyir/nodes/node_test.py]: Changed it to that order and added theq uotes.

**Comment by LonelyCat124** [comment]
Should be ready for a (hopefully) final look

File: src/psyclone/psyir/nodes/node.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/node.py:1512]: All good now, thanks.

**Pull request merged**